<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Retrieve only the best reciprocal BLAST hit for each query — blast_best_reciprocal_hit • metablastr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Retrieve only the best reciprocal BLAST hit for each query — blast_best_reciprocal_hit" />
<meta property="og:description" content="This function performs a BLAST search between query and subject sequences reciprocally and
returns (for each direction) only the best hit based on the following criteria.
A best blast hit is defined as:
the hit with the smallest e-value
if e-values are identical then the hit with the longest alignment length is chosen


A best reciprocal BLAST hit is defined as a hit that fulfills the following symmetry criteria:
blast_best_hit(A,B) = blast_best_hit(B,A)" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/installation.html">Installing metablastr and all sequence search tools</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/drostlab/metablastr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Retrieve only the best reciprocal BLAST hit for each query</h1>
    <small class="dont-index">Source: <a href='https://github.com/drostlab/metablastr/blob/master/R/blast_best_reciprocal_hit.R'><code>R/blast_best_reciprocal_hit.R</code></a></small>
    <div class="hidden name"><code>blast_best_reciprocal_hit.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs a BLAST search between query and subject sequences reciprocally and
returns (for each direction) only the best hit based on the following criteria.</p>
<p>A best blast hit is defined as:</p><ul>
<li><p>the hit with the smallest e-value</p></li>
<li><p>if e-values are identical then the hit with the longest alignment length is chosen</p></li>
</ul>

<p>A best reciprocal BLAST hit is defined as a hit that fulfills the following symmetry criteria:</p>
<p><code>blast_best_hit(A,B) = blast_best_hit(B,A)</code></p>
    </div>

    <pre class="usage"><span class='fu'>blast_best_reciprocal_hit</span><span class='op'>(</span>
  <span class='va'>query</span>,
  <span class='va'>subject</span>,
  search_type <span class='op'>=</span> <span class='st'>"nucleotide_to_nucleotide"</span>,
  strand <span class='op'>=</span> <span class='st'>"both"</span>,
  output.path <span class='op'>=</span> <span class='cn'>NULL</span>,
  is.subject.db <span class='op'>=</span> <span class='cn'>FALSE</span>,
  task <span class='op'>=</span> <span class='st'>"blastn"</span>,
  db.import <span class='op'>=</span> <span class='cn'>FALSE</span>,
  postgres.user <span class='op'>=</span> <span class='cn'>NULL</span>,
  evalue <span class='op'>=</span> <span class='fl'>0.001</span>,
  out.format <span class='op'>=</span> <span class='st'>"csv"</span>,
  cores <span class='op'>=</span> <span class='fl'>1</span>,
  max.target.seqs <span class='op'>=</span> <span class='fl'>10000</span>,
  db.soft.mask <span class='op'>=</span> <span class='cn'>FALSE</span>,
  db.hard.mask <span class='op'>=</span> <span class='cn'>FALSE</span>,
  blast.path <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query</th>
      <td><p>path to input file in fasta format.</p></td>
    </tr>
    <tr>
      <th>subject</th>
      <td><p>path to subject file in fasta format or blast-able database.</p></td>
    </tr>
    <tr>
      <th>search_type</th>
      <td><p>type of query and subject sequences that will be compared via BLAST search.
Options are:</p><ul>
<li><p><code>search_type = "nucleotide_to_nucleotide"</code></p></li>
<li><p><code>search_type = "nucleotide_to_protein"</code></p></li>
<li><p><code>search_type = "protein_to_nucleotide"</code></p></li>
<li><p><code>search_type = "protein_to_protein"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>strand</th>
      <td><p>Query DNA strand(s) to search against database/subject.
Options are:</p><ul>
<li><p><code>strand = "both"</code> (Default): query against both DNA strands.</p></li>
<li><p><code>strand = "minus"</code> : query against minus DNA strand.</p></li>
<li><p><code>strand = "plus"</code> : query against plus DNA strand.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>output.path</th>
      <td><p>path to folder at which BLAST output table shall be stored. 
Default is <code>output.path = NULL</code> (hence <code><a href='https://rdrr.io/r/base/getwd.html'>getwd()</a></code> is used).</p></td>
    </tr>
    <tr>
      <th>is.subject.db</th>
      <td><p>logical specifying whether or not the <code>subject</code> file is a file in fasta format (<code>is.subject.db = FALSE</code>; default)
or a blast-able database that was formatted with <code>makeblastdb</code> (<code>is.subject.db = TRUE</code>).</p></td>
    </tr>
    <tr>
      <th>task</th>
      <td><p>BLAST search task option (depending on the selected <code>search_type</code>). Options are:</p><ul>
<li><p><code>search_type = "nucleotide_to_nucleotide"</code></p><ul>
<li><p><code>task = "blastn"</code> : Standard nucleotide-nucleotide comparisons (default) - Traditional BLASTN requiring an exact match of 11.</p></li>
<li><p><code>task = "blastn-short"</code> : Optimized nucleotide-nucleotide comparisons for query sequences shorter than 50 nucleotides.</p></li>
<li><p><code>task = "dc-megablast"</code> : Discontiguous megablast used to find somewhat distant sequences.</p></li>
<li><p><code>task = "megablast"</code> : Traditional megablast used to find very similar (e.g., intraspecies or closely related species) sequences.</p></li>
<li><p><code>task = "rmblastn"</code></p></li>
</ul></li>
<li><p><code>search_type = "nucleotide_to_protein"</code></p><ul>
<li><p><code>task = "blastx"</code> : Standard nucleotide-protein comparisons (default).</p></li>
<li><p><code>task = "blastx-fast"</code> : Optimized nucleotide-protein comparisons.</p></li>
</ul></li>
<li><p><code>search_type = "protein_to_nucleotide"</code></p><ul>
<li><p><code>task = "tblastn"</code> : Standard protein-nucleotide comparisons (default).</p></li>
<li><p><code>task = "tblastn-fast"</code> : Optimized protein-nucleotide comparisons.</p></li>
</ul></li>
<li><p><code>search_type = "protein_to_protein"</code></p><ul>
<li><p><code>task = "blastp"</code> : Standard protein-protein comparisons (default).</p></li>
<li><p><code>task = "blast-fast"</code> : Improved BLAST searches using longer words for protein seeding.</p></li>
<li><p><code>task = "blastp-short"</code> : Optimized protein-protein comparisons for query sequences shorter than 30 residues.</p></li>
</ul></li>
</ul></td>
    </tr>
    <tr>
      <th>db.import</th>
      <td><p>shall the BLAST output be stored in a PostgresSQL database and shall a connection be established to this database? Default is <code>db.import = FALSE</code>.
In case users wish to to only generate a BLAST output file without importing it to the current R session they can specify <code>db.import = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>postgres.user</th>
      <td><p>when <code>db.import = TRUE</code> and <code>out.format = "postgres"</code> is selected, the BLAST output is imported and stored in a 
PostgresSQL database. In that case, users need to have PostgresSQL installed and initialized on their system. 
Please consult the Installation Vignette for details.</p></td>
    </tr>
    <tr>
      <th>evalue</th>
      <td><p>Expectation value (E) threshold for saving hits (default: <code>evalue = 0.001</code>).</p></td>
    </tr>
    <tr>
      <th>out.format</th>
      <td><p>a character string specifying the format of the file in which the BLAST results shall be stored.
Available options are:</p><ul>
<li><p><code>out.format = "pair"</code> : Pairwise</p></li>
<li><p><code>out.format = "qa.ident"</code> : Query-anchored showing identities</p></li>
<li><p><code>out.format = "qa.nonident"</code> : Query-anchored no identities</p></li>
<li><p><code>out.format = "fq.ident"</code> : Flat query-anchored showing identities</p></li>
<li><p><code>out.format = "fq.nonident"</code> : Flat query-anchored no identities</p></li>
<li><p><code>out.format = "xml"</code> : XML</p></li>
<li><p><code>out.format = "tab"</code> : Tabular separated file</p></li>
<li><p><code>out.format = "tab.comment"</code> : Tabular separated file with comment lines</p></li>
<li><p><code>out.format = "ASN.1.text"</code> : Seqalign (Text ASN.1)</p></li>
<li><p><code>out.format = "ASN.1.binary"</code> : Seqalign (Binary ASN.1)</p></li>
<li><p><code>out.format = "csv"</code> : Comma-separated values</p></li>
<li><p><code>out.format = "ASN.1"</code> : BLAST archive (ASN.1)</p></li>
<li><p><code>out.format = "json.seq.aln"</code> : Seqalign (JSON)</p></li>
<li><p><code>out.format = "json.blast.multi"</code> : Multiple-file BLAST JSON</p></li>
<li><p><code>out.format = "xml2.blast.multi"</code> : Multiple-file BLAST XML2</p></li>
<li><p><code>out.format = "json.blast.single"</code> : Single-file BLAST JSON</p></li>
<li><p><code>out.format = "xml2.blast.single"</code> : Single-file BLAST XML2</p></li>
<li><p><code>out.format = "SAM"</code> : Sequence Alignment/Map (SAM)</p></li>
<li><p><code>out.format = "report"</code> : Organism Report</p></li>
</ul></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores for parallel BLAST searches.</p></td>
    </tr>
    <tr>
      <th>max.target.seqs</th>
      <td><p>maximum number of aligned sequences that shall be retained. Please be aware that <code>max.target.seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></td>
    </tr>
    <tr>
      <th>db.soft.mask</th>
      <td><p>shall low complexity regions be soft masked? Default is <code>db.soft.mask = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>db.hard.mask</th>
      <td><p>shall low complexity regions be hard masked? Default is <code>db.hard.mask = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>blast.path</th>
      <td><p>path to BLAST executables.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='blast_nucleotide_to_nucleotide.html'>blast_nucleotide_to_nucleotide</a></code>, <code><a href='blast_protein_to_protein.html'>blast_protein_to_protein</a></code>,
<code><a href='blast_nucleotide_to_protein.html'>blast_nucleotide_to_protein</a></code>, <code><a href='blast_protein_to_nucleotide.html'>blast_protein_to_nucleotide</a></code>, <code><a href='blast_best_hit.html'>blast_best_hit</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>test_best_reciprocal_hit</span> <span class='op'>&lt;-</span> <span class='fu'>blast_best_reciprocal_hit</span><span class='op'>(</span>
                 query   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>'seqs/qry_nn.fa'</span>, package <span class='op'>=</span> <span class='st'>'metablastr'</span><span class='op'>)</span>,
                 subject <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>'seqs/sbj_nn_best_hit.fa'</span>, package <span class='op'>=</span> <span class='st'>'metablastr'</span><span class='op'>)</span>,
                 search_type <span class='op'>=</span> <span class='st'>"nucleotide_to_nucleotide"</span>,
                 output.path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span>,
                 db.import  <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
                 
 <span class='co'># look at results</span>
 <span class='va'>test_best_reciprocal_hit</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


