<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Count the number of motifs in a set of non-random versus randomly drawn gene promotor sequences for multiple species — motif_compare_multi_promotor_seqs • metablastr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Count the number of motifs in a set of non-random versus randomly drawn gene promotor sequences for multiple species — motif_compare_multi_promotor_seqs" />
<meta property="og:description" content="Compare the number of motifs in a set of non-random versus random promotor sequences. Internally, promotor sequences
are extracted upstream from the transcription start site (TSS) and have the length specified in interval_width.
The resulting motif count values can then be used to test the enrichment of certain motifs in real sequences compared to randomly drawn gene promotor sequences. 
Each enrichment analysis is performed for a set of different species or genomes." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/installation.html">Installing metablastr and all sequence search tools</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/drostlab/metablastr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Count the number of motifs in a set of non-random versus randomly drawn gene promotor sequences for multiple species</h1>
    <small class="dont-index">Source: <a href='https://github.com/drostlab/metablastr/blob/master/R/motif_compare_multi_promotor_seqs.R'><code>R/motif_compare_multi_promotor_seqs.R</code></a></small>
    <div class="hidden name"><code>motif_compare_multi_promotor_seqs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compare the number of motifs in a set of non-random versus random promotor sequences. Internally, promotor sequences
are extracted upstream from the transcription start site (TSS) and have the length specified in <code>interval_width</code>.
The resulting motif count values can then be used to test the enrichment of certain motifs in real sequences compared to randomly drawn gene promotor sequences. 
Each enrichment analysis is performed for a set of different species or genomes.</p>
    </div>

    <pre class="usage"><span class='fu'>motif_compare_multi_promotor_seqs</span><span class='op'>(</span>
  <span class='va'>blast_tbl</span>,
  <span class='va'>subject_genomes</span>,
  <span class='va'>annotation_files</span>,
  annotation_format <span class='op'>=</span> <span class='st'>"gff"</span>,
  <span class='va'>interval_width</span>,
  <span class='va'>motifs</span>,
  max.mismatch <span class='op'>=</span> <span class='fl'>0</span>,
  min.mismatch <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>blast_tbl</th>
      <td><p>a blast_table.</p></td>
    </tr>
    <tr>
      <th>subject_genomes</th>
      <td><p>a character vector storing the file paths to the subject genomes that shall be used as subject references.</p></td>
    </tr>
    <tr>
      <th>annotation_files</th>
      <td><p>a character vector storing the file paths to the subject annotation files in <code>.gff</code> format that match the subject genomes.</p></td>
    </tr>
    <tr>
      <th>annotation_format</th>
      <td><p>the annotation format. Options are:</p><ul>
<li><p><code>annotation_format = "gff"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>interval_width</th>
      <td><p>total number of sequences that shall be sampled per subject genome.</p></td>
    </tr>
    <tr>
      <th>motifs</th>
      <td><p>a character vector storing (case sensitive) motif sequences for which abundance in the sampled sequences shall be assessed.</p></td>
    </tr>
    <tr>
      <th>max.mismatch</th>
      <td><p>maximum number of mismatches that are allowed between the sequence motif and the matching region in the sampled sequence.</p></td>
    </tr>
    <tr>
      <th>min.mismatch</th>
      <td><p>minimum number of mismatches that are allowed between the sequence motif and the matching region in the sampled sequence.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to <code><a href='motif_compare.html'>motif_compare</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='motif_count.html'>motif_count</a></code>, <code><a href='motif_compare.html'>motif_compare</a></code>, <code><a href='motif_compare_multi.html'>motif_compare_multi</a></code>, 
<code><a href='motif_enrichment.html'>motif_enrichment</a></code>, <code><a href='motif_enrichment_multi.html'>motif_enrichment_multi</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


